File: country/trade/trade.go
package trade

import "math"

// TradeEconomy models American foreign trade, import/export flows,
// the trade deficit, tariffs, and currency strength (Dollar Index / DXY).
type TradeEconomy struct {
	AnnualExports            float64            // Total goods and services exports in USD (~$3.1T)
	AnnualImports            float64            // Total goods and services imports in USD (~$3.9T)
	TradeBalance             float64            // Net Exports = Exports - Imports (typically negative ~$800B deficit)
	DollarIndexDXY           float64            // U.S. Dollar Index (baseline 103.5)
	AverageTariffRate        float64            // Average effective tariff on imports (e.g. 0.028 for 2.8%)
	SectorExportExposure     map[string]float64 // Sector -> Export demand intensity (e.g. Tech 0.40, Energy 0.35)
	SectorImportSensitivity  map[string]float64 // Sector -> Input import sensitivity
}

// NewDefaultTradeEconomy initializes baseline US global trade dynamics.
func NewDefaultTradeEconomy() *TradeEconomy {
	exp := 3_100_000_000_000.0
	imp := 3_900_000_000_000.0

	return &TradeEconomy{
		AnnualExports:     exp,
		AnnualImports:     imp,
		TradeBalance:      exp - imp,
		DollarIndexDXY:    103.5,
		AverageTariffRate: 0.028,
		SectorExportExposure: map[string]float64{
			"Information Technology": 1.35, // High global software/semiconductor exports
			"Industrials":            1.30, // Aerospace, defense, heavy equipment
			"Energy":                 1.25, // LNG, refined petroleum exports
			"Materials":              1.10, // Chemicals, specialty materials
			"Financials":             1.15, // Global investment banking, asset management
			"Health Care":            1.05, // Pharmaceuticals
			"Consumer Discretionary": 0.90, // Autos, entertainment
			"Consumer Staples":       0.85, // Agriculture, packaged goods
			"Communication Services": 1.10, // Digital media platforms
			"Utilities":              0.50, // Domestic power
			"Real Estate":            0.40, // Domestic land/buildings
		},
		SectorImportSensitivity: map[string]float64{
			"Consumer Discretionary": 1.30, // Foreign manufactured goods, apparel, electronics
			"Information Technology": 1.20, // Foreign chip foundry inputs
			"Industrials":            1.10, // Foreign component parts
			"Materials":              1.00, // Raw mineral imports
		},
	}
}

// Tick advances global trade based on interest rates, DXY currency fluctuations, and tariffs.
func (t *TradeEconomy) Tick(domesticInterestRate float64, tickFractionOfYear float64) {
	// 1. Interest rate parity: Higher US interest rates strengthen the US Dollar (DXY)
	// Baseline Fed rate = 4.5%
	rateGap := domesticInterestRate - 0.045
	targetDXY := 103.5 + rateGap*200.0 // +100bps rate hike = +2.0 DXY points
	targetDXY = math.Max(88.0, math.Min(125.0, targetDXY))
	t.DollarIndexDXY = 0.98*t.DollarIndexDXY + 0.02*targetDXY

	// 2. Currency impact on trade:
	// A stronger dollar (DXY > 103.5) makes US exports slightly more expensive abroad,
	// and makes foreign imports cheaper into the US.
	dxyFactor := t.DollarIndexDXY / 103.5

	baselineExports := 3_100_000_000_000.0
	baselineImports := 3_900_000_000_000.0

	t.AnnualExports = baselineExports * (1.0 / math.Pow(dxyFactor, 0.4))
	t.AnnualImports = baselineImports * math.Pow(dxyFactor, 0.3) * (1.0 - t.AverageTariffRate*0.5)

	t.TradeBalance = t.AnnualExports - t.AnnualImports
}

// ExportDemandFactor returns the export multiplier for a sector.
func (t *TradeEconomy) ExportDemandFactor(sector string) float64 {
	exposure, ok := t.SectorExportExposure[sector]
	if !ok {
		exposure = 1.0
	}
	// Stronger dollar slightly dampens export multiplier, weaker dollar boosts it
	currencyCompetitiveness := 103.5 / t.DollarIndexDXY
	return 1.0 + (exposure-1.0)*0.5 + (currencyCompetitiveness-1.0)*0.3
}

File: src/country/trade.rs
use std::collections::HashMap;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TradeEconomy {
    pub annual_exports: f64,
    pub annual_imports: f64,
    pub trade_balance: f64,
    pub dollar_index_dxy: f64,
    pub average_tariff_rate: f64,
    pub sector_export_exposure: HashMap<String, f64>,
    pub sector_import_sensitivity: HashMap<String, f64>,
}

impl Default for TradeEconomy {
    fn default() -> Self {
        let exp = 2_800_000_000.0;
        let imp = 3_100_000_000.0;

        let mut export_exp = HashMap::new();
        export_exp.insert("Information Technology".to_string(), 1.30);
        export_exp.insert("Industrials".to_string(), 1.25);
        export_exp.insert("Energy".to_string(), 1.20);
        export_exp.insert("Materials".to_string(), 1.15);
        export_exp.insert("Financials".to_string(), 1.10);
        export_exp.insert("Health Care".to_string(), 1.05);
        export_exp.insert("Consumer Discretionary".to_string(), 0.90);
        export_exp.insert("Consumer Staples".to_string(), 0.85);
        export_exp.insert("Communication Services".to_string(), 1.10);
        export_exp.insert("Utilities".to_string(), 0.50);
        export_exp.insert("Real Estate".to_string(), 0.40);

        let mut import_sens = HashMap::new();
        import_sens.insert("Consumer Discretionary".to_string(), 1.30);
        import_sens.insert("Information Technology".to_string(), 1.20);
        import_sens.insert("Industrials".to_string(), 1.10);
        import_sens.insert("Materials".to_string(), 1.00);

        Self {
            annual_exports: exp,
            annual_imports: imp,
            trade_balance: exp - imp,
            dollar_index_dxy: 100.0,
            average_tariff_rate: 0.028,
            sector_export_exposure: export_exp,
            sector_import_sensitivity: import_sens,
        }
    }
}

impl TradeEconomy {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn tick(&mut self, domestic_interest_rate: f64, _tick_fraction_of_year: f64) {
        let rate_gap = domestic_interest_rate - 0.045;
        let target_dxy = (100.0 + rate_gap * 150.0).clamp(85.0, 120.0);
        self.dollar_index_dxy = 0.98 * self.dollar_index_dxy + 0.02 * target_dxy;

        let dxy_factor = self.dollar_index_dxy / 100.0;
        let baseline_exports = 2_800_000_000.0;
        let baseline_imports = 3_100_000_000.0;

        self.annual_exports = baseline_exports * (1.0 / dxy_factor.powf(0.4));
        self.annual_imports = baseline_imports * dxy_factor.powf(0.3) * (1.0 - self.average_tariff_rate * 0.5);
        self.trade_balance = self.annual_exports - self.annual_imports;
    }

    pub fn export_demand_factor(&self, sector: &str) -> f64 {
        let exposure = self.sector_export_exposure.get(sector).copied().unwrap_or(1.0);
        let currency_competitiveness = 100.0 / self.dollar_index_dxy;
        1.0 + (exposure - 1.0) * 0.5 + (currency_competitiveness - 1.0) * 0.3
    }
}
